### PR TITLE
[Refactor] Refactor  structure of `cylinder3d`

### PR DESCRIPTION
**Pull request by Xiangxu-0103**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet3d/models/voxel_encoders/voxel_encoder.py [hunk 1]
```diff
@@ -630,11 +630,14 @@ def forward(self, features: Tensor, coors: Tensor, *args,
...
         if self.compression_layers is not None:
             voxel_feats = self.compression_layers(voxel_feats)
 
-        if self.return_point_feats:
-            return voxel_feats, voxel_coors, point_feats
-        return voxel_feats, voxel_coors
+        voxel_dict['voxel_feats'] = voxel_feats
+        voxel_dict['voxel_coors'] = voxel_coors
+        voxel_dict['point2voxel_maps'] = point2voxel_maps
+        voxel_dict['point_feats'] = point_feats
```
Comment (sunjiahao1999) [inline mmdet3d/models/voxel_encoders/voxel_encoder.py:642]: It would be confusing, if we put `point_feats` into `voxel_dict`.

File: mmdet3d/models/decode_heads/cylinder3d_head.py [hunk 2]
```diff
@@ -86,72 +45,91 @@ def build_conv_seg(self, channels: int, num_classes: int,
             padding=1,
             bias=True)
 
-    def forward(self, sparse_voxels: SparseConvTensor) -> SparseConvTensor:
+    def forward(self, voxel_dict: dict) -> dict:
         """Forward function."""
-        sparse_logits = self.cls_seg(sparse_voxels)
-        return sparse_logits
+        sparse_logits = self.cls_seg(voxel_dict['voxel_feats'])
+        voxel_dict['logits'] = sparse_logits.features
```
Comment (sunjiahao1999) [inline mmdet3d/models/decode_heads/cylinder3d_head.py:51]: Should we put `logits` into `voxel_dict`?

File: mmdet3d/models/segmentors/cylinder3d.py [hunk 3]
```diff
@@ -67,16 +67,16 @@ def __init__(self,
...
 
-    def extract_feat(self, batch_inputs: dict) -> Tensor:
+    def extract_feat(self, batch_inputs_dict: dict) -> dict:
         """Extract features from points."""
-        encoded_feats = self.voxel_encoder(batch_inputs['voxels']['voxels'],
-                                           batch_inputs['voxels']['coors'])
-        batch_inputs['voxels']['voxel_coors'] = encoded_feats[1]
-        x = self.backbone(encoded_feats[0], encoded_feats[1],
-                          len(batch_inputs['points']))
+        voxel_dict = batch_inputs_dict['voxels'].copy()
```
[the remark this answers is not in the document]
Comment (Xiangxu-0103) [reply mmdet3d/models/segmentors/cylinder3d.py:72]: We should use deep copy here to avoid information coverage in `semi-supervised` learning.

File: mmdet3d/models/decode_heads/cylinder3d_head.py (hunk 2 above)
Comment (Xiangxu-0103) [reply mmdet3d/models/decode_heads/cylinder3d_head.py:51]: We will use 'inverse_map' and some other information in `voxel_dict` to map `voxel_logits` to `point_logits` in `predict` procedure. It is better to return `voxel_dict` here rather than `logits`.

File: mmdet3d/models/data_preprocessors/voxelize.py [hunk 4]
```diff
@@ -218,15 +219,13 @@ def forward(ctx: Any,
         ctx.save_for_backward(feats, voxel_feats, point2voxel_map,
                               voxel_points_count)
         ctx.mark_non_differentiable(voxel_coors)
-        if return_map:
-            return voxel_feats, voxel_coors, point2voxel_map
-        else:
-            return voxel_feats, voxel_coors
+        return voxel_feats, voxel_coors, point2voxel_map
```
[the remark this answers is not in the document]
Comment (Xiangxu-0103) [reply mmdet3d/models/data_preprocessors/voxelize.py:222]: Original implementation will cause errors during backward procedure if we set `return_map=True`. This allows `DynamicScatter3D` always returns `point2voxelmap`.

File: mmdet3d/models/voxel_encoders/voxel_encoder.py (hunk 1 above)
Comment (Xiangxu-0103) [reply mmdet3d/models/voxel_encoders/voxel_encoder.py:642]: `point_feats` should be put in `voxel_dict` for potential `PointRefineModule`.

File: mmdet3d/models/data_preprocessors/voxelize.py [hunk 5]
```diff
@@ -295,27 +295,29 @@ def forward(self, points: torch.Tensor,
                 multi-dim voxel index) of each points.
 
         Returns:
-            tuple[torch.Tensor]: A tuple contains two elements. The first one
+            tuple[torch.Tensor]: A tuple contains first elements. The first one
```
Comment (ZCMax) [inline mmdet3d/models/data_preprocessors/voxelize.py:298]: Check the 'first' representation.

File: mmdet3d/models/data_preprocessors/voxelize.py [hunk 6]
```diff
@@ -295,27 +295,29 @@ def forward(self, points: torch.Tensor,
...
             is the voxel features with shape [M, C] which are respectively
             reduced from input features that share the same voxel coordinates.
-            The second is voxel coordinates with shape [M, ndim].
+            The second is voxel coordinates with shape [M, ndim]. The third is
+            a list of point2voxel_map.
```
Comment (ZCMax) [inline mmdet3d/models/data_preprocessors/voxelize.py:302]: with shape [M]

File: mmdet3d/models/decode_heads/cylinder3d_head.py [hunk 7]
```diff
@@ -19,61 +22,17 @@ class Cylinder3DHead(Base3DDecodeHead):
...
-                     class_weight=None,
-                     loss_weight=1.0).
-        loss_lovasz (dict or :obj:`ConfigDict`): Config of Lovasz loss.
-            Defaults to dict(type='LovaszLoss', loss_weight=1.0).
-        conv_seg_kernel_size (int): The kernel size used in conv_seg.
-            Defaults to 3.
-        ignore_index (int): The label index to be ignored. When using masked
-            BCE loss, ignore_index should be set to None. Defaults to 19.
-        init_cfg (dict or :obj:`ConfigDict` or list[dict or :obj:`ConfigDict`],
-            optional): Initialization config dict. Defaults to None.
```
Comment (ZCMax) [inline mmdet3d/models/decode_heads/cylinder3d_head.py:44]: Why do need to remove these parameters in Cylinder3Dhead?

File: mmdet3d/models/decode_heads/cylinder3d_head.py (hunk 2 above)
Comment (ZCMax) [reply mmdet3d/models/decode_heads/cylinder3d_head.py:51]: If we want to put this into `voxel_dict`, maybe a comment is needed

File: mmdet3d/models/segmentors/cylinder3d.py [hunk 8]
```diff
@@ -135,10 +141,32 @@ def predict(self,
...
+                 batch_data_samples: OptSampleList = None) -> dict:
+        """Network forward process.
+
+        Args:
+            batch_inputs_dict (dict): Input sample dict which includes 'points'
+                and 'imgs' keys.
+
+                - points (List[Tensor]): Point cloud of each sample.
+                - imgs (Tensor, optional): Image tensor has shape (B, C, H, W).
+            batch_data_samples (List[:obj:`Det3DDataSample`]): The det3d data
```
Comment (ZCMax) [inline mmdet3d/models/segmentors/cylinder3d.py:163]: Defaults to None

File: mmdet3d/models/voxel_encoders/voxel_encoder.py (hunk 1 above)
Comment (ZCMax) [reply mmdet3d/models/voxel_encoders/voxel_encoder.py:642]: From my perspective, `point_feats` should not be added into `voxel_dict`, or maybe we can use `feat_dict`

File: mmdet3d/models/decode_heads/cylinder3d_head.py (hunk 7 above)
Comment (Xiangxu-0103) [reply mmdet3d/models/decode_heads/cylinder3d_head.py:44]: These parameters are implemented in `Base3DDecodeHead` and we can simplify them with `**kwargs`.

File: mmdet3d/models/segmentors/cylinder3d.py [hunk 9]
```diff
@@ -135,10 +142,32 @@ def predict(self,
         # 3D segmentation requires per-point prediction, so it's impossible
         # to use down-sampling to get a batch of scenes with same num_points
         # therefore, we only support testing one scene every time
-        x = self.extract_feat(batch_inputs_dict)
-        seg_logits_list = self.decode_head.predict(x, batch_inputs_dict,
+        voxel_dict = self.extract_feat(batch_inputs_dict)
```
Comment (Tai-Wang) [inline mmdet3d/models/segmentors/cylinder3d.py:145]: feat_dict
